### ISSUE TITLE
Enzyme don't provide support for multiple layer in component?

### ISSUE THREAD

**Issue by drd-2rcl** [enzymejs/enzyme#2235]
Hello guys, after of all, thanks for the great job!

I have a little problem when i need testing functions after add lib for translate. I reproduce code and console.log for clear the idea.

**Code working:**
```
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class AddAccesses extends Component {
  state = { isAccessEnabled: false }

  toggleInputAccess = () => {
    const { isAccessEnabled } = this.state;

    this.setState({
      isAccessEnabled: !isAccessEnabled
    });
  }

  render() {
    const { isAccessEnabled } = this.state;
    const { t } = this.props;

    return (
      <div>
        <Box
          data-testid="BoxAccess"
          bg={isAccessEnabled}
          onClick={this.toggleInputAccess}
        >
          {t('add_accesses_button_search_role_guide_off')}
        </Box>
        <Tooltip span text={t('add_accesses_tooltip_role_guide_off')} />
      </div>
    );
  }
}

AddAccesses.propTypes = {
  t: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  systems: state.systems,
  accesses: state.accesses,
});

export default connect(
  mapStateToProps, { fetchIdentities },
)(AddAccesses);

```

**Test working:**

```
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = shallow(
      <AddAccessesCaseEnzyme store={store} />,
    ).childAt(0).dive();
    instance = wrapper.instance();
  });

  it('should find BoxAccess, BoxIdentity, BoxCriticalComponent, BoxAccessKit', () => {
    console.log(wrapper.debug());
    expect(wrapper.find('[data-testid="BoxAccess"]')).toHaveLength(1);
  });
```
**Console:**
```
 <Box data-testid="BoxAccess" bg={isAccessEnabled} onClick={this.toggleInputAccess}>{t('add_accesses_button_search_role_guide_off')}</Box>
          |                                                                                                 ^
      192 |             <Tooltip span text={t('add_accesses_tooltip_role_guide_off')} />
```

So far, so good. But when i add another layer for translate in component stop work the test.

**Code working with `import { withNamespaces } from 'react-i18next';`:**
```
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withNamespaces } from 'react-i18next';

class AddAccesses extends Component {
  state = { isAccessEnabled: false }

  toggleInputAccess = () => {
    const { isAccessEnabled } = this.state;

    this.setState({
      isAccessEnabled: !isAccessEnabled
    });
  }

  render() {
    const { isAccessEnabled } = this.state;
    const { t } = this.props;

    return (
      <div>
        <Box
          data-testid="BoxAccess"
          bg={isAccessEnabled}
          onClick={this.toggleInputAccess}
        >
          {t('add_accesses_button_search_role_guide_off')}
        </Box>
        <Tooltip span text={t('add_accesses_tooltip_role_guide_off')} />
      </div>
    );
  }
}

AddAccesses.propTypes = {
  t: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  systems: state.systems,
  accesses: state.accesses,
});

export default withNamespaces()(connect(
  mapStateToProps, { fetchIdentities },
)(AddAccesses));

```

**Test with `shallow()` don't work:**

```
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = shallow(
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <BrowserRouter>
            <Switch>
              <AddAccessesCaseEnzyme />
            </Switch>
          </BrowserRouter>
        </I18nextProvider>
      </Provider>,
    ).childAt(0).dive();
    instance = wrapper.instance();
  });

  it('should user does not selected benefited', () => {
    history.push('/benefited');

    expect(window.location.href).toBe('http://localhost/benefited');
  });

  it('should find BoxAccess, BoxIdentity, BoxCriticalComponent, BoxAccessKit', () => {
    expect(wrapper.find('[data-testid="BoxAccess"]')).toHaveLength(3);
  });

  it('toggleInputIdentities and enableInputIdentities function should update state as expected', () => {
    console.log(wrapper.debug());
    instance.toggleInputAccess();
  });
```

**Console:**

```
 FAIL  app/javascript/src/spec/javascript/packs/components/organisms/add_accesses_case_enzyme.spec.js
  Testing AddAccesses component
    ✓ should user does not selected benefited (144ms)
    ✕ should find BoxAccess (10ms)
    ✕ toggleInputAccess function should update state as expected (13ms)

  ● Testing AddAccesses component › should find BoxAccess

    expect(received).toHaveLength(expected)

    Expected length: 3
    Received length: 0
    Received object: {}

      100 |
      101 |   it('should find BoxAccess', () => {
    > 102 |     expect(wrapper.find('[data-testid="BoxAccess"]')).toHaveLength(3);
          |                                                       ^
      103 |   });
      104 |
      105 |   it('toggleInputAccess function should update state as expected', () => {

      at Object.toHaveLength (app/javascript/src/spec/javascript/packs/components/organisms/add_accesses_case_enzyme.spec.js:102:55)

  ● Testing AddAccesses component › toggleInputAccess function should update state as expected

    TypeError: instance.toggleInputAccess is not a function

      105 |   it('toggleInputAccess function should update state as expected', () => {
      106 |     console.log(wrapper.debug());
    > 107 |     instance.toggleInputAccess();
          |              ^
      108 |     expect(instance.state.isAccessEnabled).toBe(true);
      109 |   });
      110 |

      at Object.toggleInputAccess (app/javascript/src/spec/javascript/packs/components/organisms/add_accesses_case_enzyme.spec.js:107:14)

  console.log app/javascript/src/spec/javascript/packs/components/organisms/add_accesses_case_enzyme.spec.js:106
    <ContextProvider value={{...}}>
      <BrowserRouter>
        <Switch>
          <LoadNamespace(Connect(AddAccesses)) />
        </Switch>
      </BrowserRouter>
    </ContextProvider>

```
**Test with `mount()` don't work:**
```
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <BrowserRouter>
            <Switch>
              <AddAccessesCaseEnzyme />
            </Switch>
          </BrowserRouter>
        </I18nextProvider>
      </Provider>,
    );
    instance = wrapper.instance();
  });

  it('should user does not selected benefited', () => {
    history.push('/benefited');

    expect(window.location.href).toBe('http://localhost/benefited');
  });

  it('should find BoxAccess', () => {
    expect(wrapper.find('[data-testid="BoxAccess"]')).toHaveLength(3);
  });

  it('toggleInputAccess function should update state as expected', () => {
    console.log(wrapper.debug());
    instance.toggleInputAccess();
    expect(instance.state.isAccessEnabled).toBe(true);
  });
```

**Console:**
```
  <Provider store={{...}}>
      <I18nextProvider i18n={{...}}>
        <BrowserRouter>
          <Router history={{...}}>
            <Switch>
              <LoadNamespace(Connect(AddAccesses)) location={{...}} computedMatch={{...}}>
                <WithMergedOptions i18n={{...}} defaultNS={[undefined]} reportNS={[undefined]} lng="pt" t={[Function: bound t]} location={{...}} computedMatch={{...}}>
                  <LoadNamespace i18n={{...}} t={[Function: bound t]} lng="pt" i18nOptions={{...}} defaultNS={[undefined]} reportNS={[undefined]} location={{...}} computedMatch={{...}}>
                    <WithMergedOptions(NamespacesConsumerComponent) ns={[undefined]} i18n={{...}} t={[Function: bound t]} lng="pt" i18nOptions={{...}} defaultNS={[undefined]} reportNS={[undefined]} location={{...}} computedMatch={{...}}>
                      <WithMergedOptions i18n={{...}} defaultNS={[undefined]} reportNS={[undefined]} lng="pt" t={[Function: bound t]} ns={[undefined]} i18nOptions={{...}} location={{...}} computedMatch={{...}}>
                        <NamespacesConsumerComponent i18n={{...}} t={[Function: bound t]} lng="pt" i18nOptions={{...}} defaultNS={[undefined]} reportNS={[undefined]} ns={[undefined]} location={{...}} computedMatch={{...}}>
                          <Connect(AddAccesses) tReady={true} i18n={{...}} t={[Function: fixedT]} lng="pt" i18nOptions={{...}} defaultNS={[undefined]} reportNS={[undefined]} location={{...}}
computedMatch={{...}}>
                            <AddAccesses tReady={true} i18n={{...}} t={[Function: fixedT]} lng="pt" i18nOptions={{...}} defaultNS={[undefined]} reportNS={[undefined]} location={{...}} computedMatch={{...}} systems={{...}} accesses={{...}} selectedBenefited={{...}} fetchIdentities={[Function]} fetchCriticalComponents={[Function]} fetchSystems={[Function]} setSelectedAccesses={[Function]} updateSelectedAccesses={[Function]}>
                              <div>
                                <styled.div borderBottom={true} alignItems={true}>
                                  <StyledComponent borderBottom={true} alignItems={true} forwardedComponent={{...}} forwardedRef={{...}}>
                                    <div className="sc-dxgOiQ KKGvM">
                                      <Link to="/helpdesk">
                                        <a onClick={[Function: onClick]} href="/helpdesk">
                                          <styled.img src="test-file-stub">
                                            <StyledComponent src="test-file-stub" forwardedComponent={{...}} forwardedRef={{...}}>
                                              <img src="test-file-stub" className="sc-ckVGcZ dvqRth" />
                                            </StyledComponent>
                                          </styled.img>
                                        </a>
                                      </Link>
                                    </div>
                                  </StyledComponent>
                                </styled.div>
                                <styled.div spaceAround={true}>
                                  <StyledComponent spaceAround={true} forwardedComponent={{...}} forwardedRef={{...}}>
                                    <div className="sc-dxgOiQ gfxngO">
                                      <styled.div flexEnd={true}>
                                        <StyledComponent flexEnd={true} forwardedComponent={{...}} forwardedRef={{...}}>
                                          <div className="sc-dxgOiQ kgTxfX">
                                            <styled.button data-testid="BoxAccess" bg={{...}} onClick={[Function]}>
                                              <StyledComponent data-testid="BoxAccess" bg={{...}} onClick={[Function]} forwardedComponent={{...}} forwardedRef={{...}}>
                                                <button data-testid="BoxAccess" onClick={[Function]} className="sc-chPdSV gDtCiS">
                                                  Eu sei o nome ou parte do nome do perfil
                                                </button>
```
**In short, when i add lib `i18next-react` the test gets lost, and i don't get access for execute the function**

```
 <button data-testid="BoxAccess" onClick={[Function]} className="sc-chPdSV gDtCiS">
                                                  Eu sei o nome ou parte do nome do perfil
                                                </button>
```

**Comment by DanielFerrariR** [comment]
@ljharb Hello, help us please.

**Comment by ljharb** [comment]
@drd-2rcl Instead of wrapping the component with your provider, use the `wrappingComponent` option, like this:
```js
function WrapWithProviders({ children }) {
  return (
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <BrowserRouter>
          <Switch>
            {children}
          </Switch>
        </BrowserRouter>
      </I18nextProvider>
    </Provider>
  );
}
const wrapper = shallow(<AddAccessesCaseEnzyme />, {
  wrappingComponent: WrapWithProviders,
});
```

**Comment by daiky00** [comment]
@ljharb did this answer you provide change?

**Comment by jakerobb** [comment]
I'm going to resurrect this old issue to reassert @daiky00's question -- @ljharb, did this change? 

I'm using enzyme 3.11.0, and when I try the suggestion to use your `wrappingComponent` suggestion above, I get this error:

```
 Method “type” is meant to be run on 1 node. 2 found instead.
```

`2` does not seem to be the number of wrapping providers -- I get this exact message even when there's only one provider in `WrapWithProviders`.

**Comment by ljharb** [comment]
I don't believe it should have changed at all.

@jakerobb what does your component and test code look like?

**Comment by jakerobb** [comment]
After a few hours of spelunking with the debugger, I figured out my problem. I was reusing a wrapper component from our production app which looks like this:

```
import React from 'react';
import defaultTheme from './defaultTheme';
import { CssBaseline } from '@material-ui/core';
import { ThemeProvider as MUIThemeProvider, StylesProvider, createGenerateClassName } from '@material-ui/core/styles';
interface Props {
    children: React.ReactNode;
}

const ThemeProvider = ({ children }: Props) => {
    const generateClassName = createGenerateClassName({
        seed: 'a',
    });

    return (
        <StylesProvider generateClassName={generateClassName}>
            <MUIThemeProvider theme={defaultTheme}>
                <CssBaseline />
                {children}
            </MUIThemeProvider>
        </StylesProvider>
    );
};

export default ThemeProvider;
```

The two child elements inside `MUIThemeProvider` were the two nodes in question. For purposes of my test, I created an almost-identical provider component which omits `CssBaseline`, and I'm in business.

Thanks, and sorry for the ping. If you have time to explain, I'd love to have a clearer understanding of why it doesn't/can't work the way it was.

**Comment by ljharb** [comment]
Specifically, it’s because enzyme is trying to locate a single child produced by the provider (which is how providers usually operate) so it can be used as the primary component enzyme is wrapping. It wouldn’t make sense to wrap more than one, and there’s no easy way to know which child you’re intending.

**Comment by ljharb** [comment]
Closing as answered.
